tools/swift-inspect/Sources/SwiftInspectLinux/PTrace.swift

```diff
@@ -0,0 +1,157 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+import LinuxSystemHeaders
+
+// Provides scoped access to a PTrace object.
+public func withPTracedProcess(pid: pid_t, _ closure: (consuming PTrace) throws -> Void) throws {
+  let ptrace = try PTrace(pid)
+  try closure(ptrace)
+}
+
+public struct PTrace: ~Copyable {
+  enum PTraceError: Error {
+    case operationFailure(_ command: CInt, pid: pid_t, errno: CInt = get_errno())
+    case waitFailure(pid: pid_t, errno: CInt = get_errno())
+    case unexpectedWaitStatus(pid: pid_t, status: CInt, sigInfo: siginfo_t? = nil)
+  }
+
+  let pid: pid_t
+
+  // Initializing a PTrace instance attaches to the target process, waits for
+  // it to stop, and leaves it in a stopped state. The caller may resume the
+  // process by calling cont().
+  // NOTE: clients must use withPTracedProcess instead of direct initialization.
+  init(_ pid: pid_t) throws {
```

Comment:
You can mark this as `fileprivate` to ensure that no one else is able to construct the type without the `withPTracedProcess`.